Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -0,0 +1,341 @@
+    }
+
+    const files: string[] = findFiles(ftpOptions);
+    let ftpClient = await getFtpClient(ftpOptions);
+    const tracker = new ProgressTracking(ftpOptions, files.length + 1); // add one extra for the root directory
+
+    let retryWithNewClient = async (task: () => {}, retry: number) => {
+        let e = null;
+        while (retry > 0) {
+            try {
+                await task();
+                return;
+            } catch (err) {
+                e = err;
+                tl.warning(err);
+                ftpClient.close();
+
+                await sleep(1000);
+                ftpClient = await getFtpClient(ftpOptions);
+
+                retry--;
+            }
+        }
+
+        if (e) {
+            throw e;
+        }
+    };
+
+    try {
+        // clean option all assumes the folder exists as it tries to cd into it
+        await retryWithNewClient(async () => {
+            await ftpClient.ensureDir(ftpOptions.remotePath);
+        }, tries);
+        if (ftpOptions.clean) {
+            console.log(tl.loc("CleanRemoteDir", ftpOptions.remotePath));
+            await retryWithNewClient(async () => {
+                await ftpClient.removeDir(ftpOptions.remotePath);
+            }, tries);
+        } else if (ftpOptions.cleanContents) {
+            console.log(tl.loc("CleanRemoteDirContents", ftpOptions.remotePath));
+            await retryWithNewClient(async () => {
+                await ftpClient.cd(ftpOptions.remotePath);
+                await ftpClient.clearWorkingDir();
+            }, tries);
+        }
+
+        // remotely, ensure we are in remote path
+        await retryWithNewClient(async () => {
+            await ftpClient.ensureDir(ftpOptions.remotePath);
+        }, tries);
+
+        for (const file of files) {
+            tl.debug("file: " + file);
+            var remoteFile: string = ftpOptions.preservePaths
+                ? path.join(
+                    ftpOptions.remotePath,
+                    file.substring(ftpOptions.rootFolder.length)
+                )
+                : path.join(ftpOptions.remotePath, path.basename(file));
+
+            remoteFile = remoteFile.replace(/\\/gi, "/"); // always use forward slashes
+            tl.debug("remoteFile: " + remoteFile);
+
+            let stats = tl.stats(file);
+            if (stats.isDirectory()) {
+                await retryWithNewClient(async () => {
+                    await ftpClient.ensureDir(remoteFile);
+                }, tries);
+                tracker.directoryProcessed(remoteFile);
+            } else if (stats.isFile()) {
+                // upload files
+                await retryWithNewClient(async () => {
+                    await ftpClient.upload(fs.createReadStream(file), remoteFile);
+                }, tries);
+                tracker.fileUploaded(file, remoteFile);
+            }
+        }
+
+        console.log(tl.loc("UploadSucceedMsg", tracker.getSuccessStatusMessage()));
```

Comment:
Do you need to set 'success' for the result?

Reply:
I don't know, I didn't, but the task was successful anyway during my tests.  I guess if we don't fail it, it defaults to `succeeded`?